# HG changeset patch
# User Rik <rik@octave.org>
# Date 1431445613 25200
#      Tue May 12 08:46:53 2015 -0700
# Node ID 89616a98b02e0f9cff77601f4ed9b5c58b755a3c
# Parent  5212cda4ac0e36cdfeed1ec5bf939cffe17892a1
regexptranslate.m: Fix 'escape' option list of special chars (bug #45084).
Also speed up 'wildcard' option by 44% by using strrep rather than regexprep.

* regexptranslate.m: Explicitly list all all special regexp chars in the
regular expression which escapes them.  Replace regexprep calls with strrep
function calls in 'wildcard' option processing.

diff --git a/scripts/strings/regexptranslate.m b/scripts/strings/regexptranslate.m
--- a/scripts/strings/regexptranslate.m
+++ b/scripts/strings/regexptranslate.m
@@ -59,30 +59,30 @@ function y = regexptranslate (op, s)
   endif
 
   if (! ischar (op))
     error ("regexptranslate: operation OP must be a string");
   endif
 
   op = tolower (op);
   if (strcmp ("wildcard", op))
-    y = regexprep (regexprep (regexprep (s, '\.', '\.'),
-                                            '\*', '.*'),
-                                            '\?', '.');
+    y = strrep (strrep (strrep (s, '.', '\.'),
+                                   '*', '.*'),
+                                   '?', '.');
   elseif (strcmp ("escape", op))
-    y = regexprep (s, '([^\w])', '\\$1');
+    y = regexprep (s, '([][(){}.*+?^$|\\])', '\\$1');
   else
     error ("regexptranslate: invalid operation OP");
   endif
 
 endfunction
 
 
 %!assert (regexptranslate ("wildcard", "/a*b?c."), "/a.*b.c\\.")
-%!assert (regexptranslate ("escape", '$.?[abc]'), '\$\.\?\[abc\]')
+%!assert (regexptranslate ("escape", '^.?[abc]$'), '\^\.\?\[abc\]\$')
 
 ## Test input validation
 %!error <Invalid call to regexptranslate> regexptranslate ()
 %!error <Invalid call to regexptranslate> regexptranslate ("wildcard")
 %!error <Invalid call to regexptranslate> regexptranslate ("a", "b", "c")
 %!error <invalid operation> regexptranslate ("foo", "abc")
 %!error <operation OP must be a string> regexptranslate (10, "abc")
 
